packages/zoltan2/src/algorithms/connectivity/Zoltan2_IceSheet.hpp

```diff
@@ -0,0 +1,318 @@
+#include<Zoltan2_AlltoAll.hpp>
+
+/*! \file Zoltan2_IceSheet.hpp
+ * \brief The user interface for ice-sheet connectivity functionality.
+ *
+ * This file contains the entry point for the ice sheet connectivity
+ * label propagation algorithm. 
+*/
+
+namespace Zoltan2{
+  template <typename Adapter>
+  class IceProp : public Algorithm<Adapter> {
+    public:
+      typedef typename Adapter::base_adapter_t base_adapter_t;
+      typedef typename Adapter::lno_t lno_t;
+      typedef typename Adapter::gno_t gno_t;
+      typedef typename Adapter::offset_t offset_t;
+      typedef typename Adapter::scalar_t scalar_t;
+      typedef typename Adapter::part_t part_t;
+      typedef typename Adapter::user_t user_t;
+      typedef typename Adapter::userCoord_t userCoord_t;
+      typedef Tpetra::Map<lno_t, gno_t> map_t;
+      
+      
+      /*! The constructor for the IceProp class
+       *  
+       *  \param problemComm A Teuchos::RCP<const Teuchos::Comm<int>> representing the communicator for the problem.
+       *  \param adapter A Teuchos::RCP<const GraphAdapter<>> representing the structure of the ice sheet mesh's bottom layer.
+       *  \param basalFriction A boolean pointer to an array of flags for each local vertex with true representing a grounded vertex and false representing a floating vertex.
+       *  \param boundary_edges An array of global vertex identifiers representing all of the edges that the current process knows about which are on the boundary between the ice and the water. 
+       *                        The IDs must be present in the overlap map of the current process if the ID is not owned by this process. Edges should not be specified more than once.
+       *  \param num_boundary_edges An integer representing the number of boundary edges on the current process. 2*num_boundary_edges is the length of the boundary_edges array.
+       *
+       **/
+      IceProp(const RCP<const Comm<int> > &problemComm__,
+	      const RCP<const GraphAdapter<user_t,userCoord_t> > &adapter__, 
+	      bool* basalFriction, gno_t* boundary_edges__, int num_boundary_edges__):
+         adapter(adapter__), grounding_flags(basalFriction), boundary_edges(boundary_edges__),
+         num_boundary_edges(num_boundary_edges__), problemComm(problemComm__)
+      {
+
+	env = rcp(new Environment(problemComm));
+        modelFlag_t flags;
+	flags.reset();
+
+	///build the graph model from the GraphAdapter.	 
+	buildModel(flags);
+      }
+   
+     /*! This function performs the connectivity check.
+      *
+      *  \return An array of flags for each local vertex. Values are be interpreted as:
+      *          -2 for a vertex that should be kept (these vertices are sufficiently connected to the ground)
+      *          -1 for a vertex that has no connection to the ground (e.g. an iceberg).
+      *          <vtxID> for a vertex that is in a floating hinge, where vtxID is the hinge vertex.
+      */
+      int* getDegenerateFeatureFlags();
```

Comment:
I think this return type should change. We make use of negative integers, so is it safe to use the global ordinal type?

Reply:
Given the documentation, you are correct that int is not appropriate because you are returning global ordinals in some cases.
Global ordinal type can also be a challenge; it is almost always signed, but Trilinos does offer Tpetra_INST_INT_UNSIGNED_LONG_LONG, which would have unsigned global ordinals.

I'm trying to think of something clever, assuming we return data as gno_t.  One option:
define constants Zoltan2::AttachedToGround = std::numeric_limits<gno_t>::max and Zoltan2::Floating = std::numeric_limits<gno_t>-1  and return those instead of -2 and -1?  That's not perfect, as it prevents the user from using these two global IDs (but we can check for that condition in the Map).

What clever ideas do you have?

Reply:
Your suggestion sounds like it would be the best way to implement the current approach with unsigned numbers in mind. With absolute scalability in mind, it seems like this approach would be problematic for cases where the number of global IDs used approaches the limit of the global ordinal type. Now, that is very unlikely to happen, but I'm not sure how many ways we'd be able to resolve that issue without sacrificing some functionality. It seems like our flags would be ambiguous to the user in that case, unless I'm misunderstanding your note about using the map to check for the global IDs, which is entirely possible.

Maybe it would be best to return two arrays, as we are currently attempting to convey two types of information with the same array. For instance, an integer array to indicate a vertex's status, where 0 = keep, 1 = floating, and 2 = hinged, and then a separate global ordinal array that indicates the hinge vertex where applicable. This is more memory-intensive, and since very few vertices are typically removed it may not be as efficient as our previous approach, but it allows all global IDs to be used, and still conveys the same information. This may also result in more readable code, as it would be clear that the different arrays indicate different things (we aren't mixing sentinel values and vertex identifiers). I'm thinking in this approach we would use just some default value (e.g. 0) for a vertex that does not have a status of 2, and <vtxID> for a vertex with status 2. One drawback is that it would be easier to misinterpret these hinge identifiers, as all of the entries are valid IDs.

Reply:
OK, if you are going to return two arrays, please use the utility interface in which the user is responsible for allocating the arrays.  It might be good for them to be Teuchos::ArrayView, then, so that we can error-check that they have sufficient size.  Similarly, we can use ArrayView for the input arguments (basal friction, boundary edges).

Reply:
Noted. I will get to work on this today.

Reply:
I changed the return type of the propagation function so that it returns the ArrayRCP of vertex labels. I am then going to interpret the labels in the utility function to fill in the two output ArrayViews.

Reply:
I don't understand. The ArrayRCP was needed if we used a class interface with getDegenerateFeatureFlags.  But with a utility interface, we aren't saving pointers to any data, so we don't need an ArrayRCP, right?

Reply:
The class that does the actual propagation stores the labels in an ArrayRCP<const VtxLabel>, so the propagation function currently returns that array of labels to the utility function. Then, the utility function will interpret the labels and set the output ArrayViews appropriately.